### PR TITLE
Prevent segfault while reading

### LINKED ISSUE #46397: Segmentation error during reading.
# Describe the bug

Segmentation error happening when reading the book for a long time.

# Steps To Reproduce

1. Load the save. 
2. Start reading the book.
3. At random period of time segmentation error happens and game crashes.

# Expected behavior

No segmentation errors.

# Screenshots

# Versions and configuration
- OS: Windows
    - OS Version: MINGW/CYGWIN/MSYS2 on unknown Windows version
- Game Version: 0.E-8531-gea3e6a3 [64-bit]
- Graphics Version: Tiles
- Game Language: English [en]
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Bionic Professions [package_bionic_professions]
]

# Additional context

The issue seemingly manifested itself after consuming gamma globulin drug in attempt to counter toxins & bringing the fresh haul of books from the school library. 

[crash.log](https://github.com/CleverRaven/Cataclysm-DDA/files/5751867/crash.log)
[debug.log](https://github.com/CleverRaven/Cataclysm-DDA/files/5751868/debug.log)
[full save.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/5753214/full.save.zip)

### PR DESCRIPTION
**Pull request by Ramza13**
#### Summary
SUMMARY: Bugfixes "Prevent segfault while reading"

#### Purpose of change
Fixes #46397
Prevent segfault while reading

#### Describe the solution
Book was being dropped while reading and attempt to access it was seg faulting.  Added check.  Since this has caused additional seg faults, added code to cancel activity upon dropping wielded item due to toxin buildup.

### CODE REVIEW

**Comment by Stoopedhooy** [comment]
Oh, wow, thank you for figuring this out!

**Comment by actual-nh** [comment]
Why was the book getting dropped, BTW?

**Comment by Ramza13** [comment]
The character has toxin_buildup which can cause you to randomly drop your wielded item.  This isn't the first segfault it's caused so I'm gonna have it cancel your activity to prevent more of these.

File: src/player_hardcoded_effects.cpp [hunk 1]
```diff
@@ -1590,6 +1590,7 @@ void Character::hardcoded_effects( effect &it )
                     } else if( limb == "hand" ) {
                         if( is_armed() && can_drop( weapon ).success() ) {
                             if( dice( 4, 4 ) > get_dex() ) {
+                                cancel_activity();  //Prevent segfaults from activities trying to access missing item
```
Comment (actual-nh) [inline src/player_hardcoded_effects.cpp:1593]: Weapon?

Comment (Ramza13) [reply src/player_hardcoded_effects.cpp:1593]: Whatever you are wielding is the characters weapon even if its a book or tool or whatever.

Comment (actual-nh) [reply src/player_hardcoded_effects.cpp:1593]: I suppose it can technically be a melee weapon in any event...

**Pull request merged**